pgx/src/pgbox.rs

```diff
@@ -80,46 +82,62 @@ use std::ops::{Deref, DerefMut};
 ///     // we can't free it
 /// }
 /// ```
-///
-///
-/// ## Safety
-///
-/// TODO:
-///  - Interatctions with Rust's panic!() macro
-///  - Interactions with Postgres' error!() macro
-///  - Boxing a null pointer -- it works ::from_pg(), ::into_pg(), and ::to_pg(), but will panic!() on all other uses
-///
-pub struct PgBox<T> {
-    inner: Inner<T>,
+#[repr(transparent)]
+pub struct PgBox<T, AllocatedBy: WhoAllocated<T> = AllocatedByPostgres> {
+    ptr: Option<NonNull<T>>,
+    __marker: PhantomData<AllocatedBy>,
 }
 
-struct Inner<T> {
-    ptr: Option<*mut T>,
-    allocated_by_pg: bool,
+pub trait WhoAllocated<T> {
+    fn free(ptr: *mut T);
```

Comment:
Interesting that `free()` doesn't take `self`

Reply:
These are, I guess you'd call them, "associated types"?  So there's no instance of them anywhere to reference as `self`.  This is how I was able to reduce this thing to the size of a pointer.